### PR TITLE
add 2d physic optimization similar to 3d physics

### LINKED ISSUE #27413: Applied torque has no effect if CollisionShape2D starts disabled
**Godot version:** v3.1.stable.official
**OS/device including version:** Windows 10 Pro

**Issue description:**
This issue was found while working on the Space Rocks example of the Chris Bradfield book.
When the CollisionShape2D starts disabled (either in the inspector or in the _ready() method), applied_torque has no effect. Debugging shows that the torque is indeed applied but the transform matrix doesn't account for it. Forces are applied normally, though. I can thrust forward, but not rotate it.

If the CollisionShape2D starts enabled and then gets disabled (by a timer of 0.01s, for example), enabling it again works as expected (I can rotate it by applying a torque).

I didn't add this interaction in the minimal example, but if another RigidBody2D collides with the player, it is pushed back (force is applied) but no rotation is applied.

The original Space Rocks source, when imported into 3.1, suffers from this aswell. In 3.0.6 it doesn't, though.

**Minimal reproduction project:**
[test.zip](https://github.com/godotengine/godot/files/3004746/test.zip)

You can either disable the CollisionShape2D in the inspector or enable it and uncomment _ready() to test both cases.

### PR DESCRIPTION
**Pull request by Raphael2048**
Add 2d physic optimization similar to 3d physics.
according to https://github.com/godotengine/godot/issues/16829
this also fix https://github.com/godotengine/godot/issues/27413

In additionally, as we have optimization outside, I think we can delete the optimization here:
https://github.com/godotengine/godot/blob/02319dceb2e17184eb765c67719a306f56dafc1b/servers/physics_2d/body_2d_sw.cpp#L36-L40
I'll update my PR if you think we should delete it.
If the optimization here is deleted, this question will be fixed too. https://github.com/godotengine/godot/issues/29888

### CODE REVIEW

File: servers/physics/collision_object_sw.h [hunk 1]
```diff
@@ -139,8 +139,11 @@ class CollisionObjectSW : public ShapeOwnerSW {
 	_FORCE_INLINE_ void set_ray_pickable(bool p_enable) { ray_pickable = p_enable; }
 	_FORCE_INLINE_ bool is_ray_pickable() const { return ray_pickable; }
 
-	_FORCE_INLINE_ void set_shape_as_disabled(int p_idx, bool p_enable) { shapes.write[p_idx].disabled = p_enable; }
-	_FORCE_INLINE_ bool is_shape_set_as_disabled(int p_idx) const { return shapes[p_idx].disabled; }
+	void set_shape_as_disabled(int p_idx, bool p_enable);
+	_FORCE_INLINE_ bool is_shape_set_as_disabled(int p_idx) const {
+		CRASH_BAD_INDEX(p_idx, shapes.size());
```
Comment (akien-mga) [inline servers/physics/collision_object_sw.h]: Why crash? `ERR_FAIL_INDEX_V` should be fine.

Comment (raphael10241024) [reply servers/physics/collision_object_sw.h]: Copied from here:
https://github.com/godotengine/godot/blob/297e7efa0af0765b3030dce726e54dde0b37f95c/servers/physics_2d/collision_object_2d_sw.h#L120-L122
It there is an error, it proves that our code is wrong. I think crash is fine.

Comment (akien-mga) [reply servers/physics/collision_object_sw.h]: Ah actually I put these `CRASH_BAD_INDEX` instead of `ERR_FAIL_INDEX_V` myself in cdc411fd54fbafc414b29abb1d50a0ff327f0d89 :D

GCC was raising a warning:
```
servers/physics_2d/collision_object_2d_sw.h:119:56: warning: returning reference to temporary [-Wreturn-local-addr]
```

For the reference, here are the logs from the discussion on IRC: http://godot.eska.me/irc-logs/meeting/2018-09-27.log
```
[16:07:09] <Akien> Those four getters trigger the warning: https://github.com/godotengine/godot/blob/master/servers/physics_2d/collision_object_2d_sw.h#L118-L133
[16:07:28] <reduz> those are weird because there is no reason for that code to return a reference to a temporary
[16:07:37] <Akien> Because they're `const Type& const` or something but the fail condition returns a non const obj
[16:07:40] <reduz> i thought with the changes tmm[m] did, they would be gone but they were not
[16:07:58] <Akien> It's only those 4 occurrences + one in gdscript that raise the warning
[16:08:30] <reduz> ah dirhg
[16:08:33] <reduz> er right
[16:08:36] <reduz> the error condition does not
[16:09:12] <reduz> I would then replace them with
[16:09:15] <reduz> CRASH_BAD_INDEX
[16:09:26] <reduz> since thre is no point in returning anything there, its not there an user accessible API to throw an error
[16:09:27] <Akien> The GDScript one is: https://github.com/godotengine/godot/blob/master/modules/gdscript/gdscript.h#L149-L152
[16:09:42] <Akien> Hm, makes sense
[16:10:01] <reduz> same for that one too, they are not user accessible
[16:11:05] <Akien> Alright, I'll do those changes and merge afterwards
```

TL;DR: All good :)

**Comment by akien-mga** [comment]
Thanks!
